New Teamspeak Component

Gerto opened the pull request:
Checks  status of teamspeak server and tracks amount of users connected.

``` yaml
  platform: teamspeak
  host: example.com
  username: serveradmin
  password: password
```

I'll write documentation as well if this gets accepted. I'm relatively inexperienced with python so would greatly appreciate if someone could double check the code.
(Closed original pull request as I completely messed up squashing commits)

Gerto committed:
New Teamspeak Component
Updated .coveragerc
Updated requirements_all.txt

balloob commented on homeassistant/components/sensor/teamspeak.py:
```diff
@@ -0,0 +1,128 @@
+
+DEFAULT_NAME = 'Teamspeak Server'
+ATTR_SERVER_NAME = 'Name'
+ATTR_MAX_USERS = 'Max. Users'
```
The attribute names need to be lowercase and not contain spaces.

Gerto replied:
Thank you, I try to look at other code for examples and find several other modules not respecting this then:
-cpuspeed
-swiss_public_transport

I cannot find a single sensor that does it another way , except maybe vera, they don't use a variable but just put it in the code

`attr['Vera Device Id'] = self.vera_device.vera_device_id`

Is that better? Looks strange?
Maybe just remove the attributes? Showing something like "max_users" in frontend doesn't seem too user friendly?

balloob replied:
It's up to you if you want to add the attributes to the sensor. Not sure if something like max users is really useful.

Attributes are allowed to have spaces and periods in them but it is bad practice. For example, printing an attribute in a template has the format `state.attributes.max_users`. As you can see, that wouldn't work with periods or spaces in the attribute name.

balloob commented on homeassistant/components/sensor/teamspeak.py:
```diff
@@ -0,0 +1,128 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Sets up the Teamspeak sensor. """
+    name = config.get('name', DEFAULT_NAME)
+    host = config.get('host', None)
```
When using `get`, `None` is the default value if none given.

balloob commented on homeassistant/components/sensor/teamspeak.py:
```diff
@@ -0,0 +1,128 @@
+    host = config.get('host', None)
+    username = config.get('username', None)
+    password = config.get('password', None)
+    virtualserver_id = config.get('virtualserver_id', 0)
```
If you expect this to be a number, consider using `convert`
```python
from homeassistant.util import convert

virtualserver_id = convert(config.get('virtualserver_id'), int, 0)
```

balloob commented on homeassistant/components/sensor/teamspeak.py:
```diff
@@ -0,0 +1,128 @@
+        """ Icon to use in the frontend, if any. """
+        return ICON
+
+    def connect(self):
```
Because you call this in the constructor of the entity, there is no way for your setup method to know if the setup was correct. Instead, inline this method in your setup method and pass the `_ts3conn` object into the constructor. That way you can avoid calling `add_devices` on an incorrect Entity.

balloob replied:
This comment still needs fixing before this can get merged.

Gerto replied:
I honestly don't understand what you mean; I thought this was fixed when I removed the attributes. The quote is a large block of code, but I'll assume you're talking about the connect function

After reading your comment 50 times, I think you mean I need to add _ts3conn as another argument for __init__ and then maybe try to connect before calling add_devices in setup_platform somehow?

balloob replied:
You are returning `False` from this method when the connection fails but you're not using this piece of information anywhere. So if you would put in wrong information in your component, the connection will fail but the component will act as if it succeeded and add it to Home Assistant. Home Assistant will poll this every 15 seconds returning in an error in the logs every time.

See for an example the [forecase.io component](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/sensor/forecast.py#L54-L63).

Gerto committed:
Updated teamspeak.py
Some improvements
Removed attribute and extra info functionality

Gerto committed:
Merge pull request #1 from Gerto/fix

Updated teamspeak.py

flyte commented:
Good work on this integration @Gerto - I'm looking forward to having these stats in my HA :+1:

balloob commented on homeassistant/components/sensor/teamspeak.py:
```diff
@@ -0,0 +1,112 @@
+    @property
+    def name(self):
+        """ The name of the sensor. """
+        return self._name
```
Why not use the server name as name ?

Gerto replied:
Thank you for the suggestion; I think the user should have the option to set it in config, but if name isn't set it defaults to the TS server name. For example, the server I'm monitoring has a long name and this looks bad in Home Assistant.

So, I added a check in init (this seems best because name should only be set at initialization and doesn't change?) to see if config name is none
However, I'm a bit stuck with how exactly to write this line:
`self._name = self._ts3conn.serverlist()[self._virtual_server_id]['virtualserver_name']`
is too long for lynting, so how do I split this?

1)Seems ridiculous?
```
        self._name = self._ts3conn.serverlist()[
            self._virtual_server_id
            ][
                'virtualserver_name'
            ]
```
2)This also works, but is using many variables really better than just a long line?
```
            serverlist = self._ts3conn.serverlist()
            virtualserver = serverlist[self._virtual_server_id]
            self._name = virtualserver['virtualserver_name']
```

Thank you

Gerto replied:
Good idea,ty; I've changed it now so that the user can still set it in config but it will default to the server name. I've done this in the init function because it seems the name only has to be set at initialization? Or is this bad practice?

Gerto committed:
Changed Name

When name is not set in config, component will now use the TS server name as default.

balloob commented:
Closing this PR as it seems to have gone stale. You're welcome to re-open it when the comments have been addressed.

Gerto commented:
gone stale?????
I'm just trying to help but keep getting increasingly impossible tasks, even though many other components have the exact same code that is supposedly bad but that's ignored.

I just wanted to help and thought this'd be a nice addition, for anyone that needs it, the component is working perfectly so you can just add it.

balloob commented:
There is a clear bug. If you put in broken info you will keep getting errors every 30 seconds. I gave this comment over a month ago and it has not been addressed nor does there seem to be any movement. I call this gone stale.

If other components have the same style of broken code (and I doubt that) it should be fixed and is never an argument to put in more broken code.
